src/librustc/hir/mod.rs

```diff
@@ -1723,6 +1723,12 @@ pub enum TyKind {
     ///
     /// Type parameters may be stored in each `PathSegment`.
     Path(QPath),
+    /// A type definition itself. This is currently only used for the `existential type`
+    /// item that `impl Trait` in return position desugars to.
+    ///
+    /// The generic arg list are the lifetimes (and in the future possibly parameters) that are
+    /// actually bound on the `impl Trait`.
+    Def(ItemId, HirVec<GenericArg>),
```

Comment:
I want something like this in general, maybe `{...}` block syntax for types?
Then `-> impl Trait` would be `-> { existential type X: Trait; X }`.
cc @Centril @nikomatsakis

Reply:
In other words this would be legal?
```rust
type Foo = { u8 };
```
Not sure why it is useful tho...

Reply:
isn't that essentially generic modules? (but unnamed)

Reply:
@oli-obk I guess only if you can depend on parameters in scope.

@Centril for the same reason we have it for expressions :P.

Reply:
> I guess only if you can depend on parameters in scope.

which impl trait always does

Reply:
@eddyb in exprs it exists for some sort of encapsulation purposes... I suppose you could do some sort of encapsulation of type aliases... but we should tread carefully here, there might be ambiguity in store.

PS: we should move this conversation elsewhere into a fresh issue / internals instead of comments on a merged PR (totally not discoverable...).

Reply:
@Centril 
>  I suppose you could do some sort of encapsulation of type aliases

Module-based encapsulation is how `existential type X: Trait;` works today.

Reply:
@cramertj I meant something like this:

```rust
type Foo = {
    type ComplexType = Bar<Baz<u8>>;
    Vec<ComplexType>
};
```